### ISSUE TITLE
NOx ensemble member run with offline chemistry

### ISSUE THREAD

**Issue by cschooling** [geoschem/geos-chem#2804]
### Your name

Chloe Schooling

### Your affiliation

The University of Edinburgh 

### Please provide a clear and concise description of your question or discussion topic.

Hi, I have added some additions to the carbon run source code, in order to add the chemical species NOx. 

This has worked well and gets reasonable agreement in comparison to Full Chemistry model runs. 
The changes I have made to KPP include adding in the additional reactions in carbon.eqn:
```
NOx             = Dummy             : k_Trop(4);
DummytoNOx      = NOx               : k_Trop(5);
```
Then I determine the chemical rates within carbon_Funcs.F90: 
```
 loss = LNOx_chem*C(ind_NOx)/NOx_lev
 prod = PNOx_chem*C(ind_NOx)/NOx_lev
 k_Trop(4) = loss   /C(ind_NOx)
 k_Trop(5) = prod 
```
Where LNOx_chem, PNOx_chem, and NOx_lev are read in using HEMCO (they are rates, and concentrations from a full chemistry run). As I found that scaling the nox chemical rates by relative differences in the nox concentration compared to a baseline full chem run is a pretty good approximation.  I also integrated the nox species into the carbon_gases_mod.F90 script within GeosCore. 

All of this works to model and investigate transport/emissions of atmiospheric NOx more efficiently than turning on the full chemistry mechanism. 

The reason I have done this is in order to explore inversion work to improve NOx emissions estimates. I am hoping to run this using an ensemble kalman filter (ENKF) approach. This involves running the simulation with many different NOx ensemble members all with different emission scaling factors. I am trying to run with 100 ensemble members, which I label NOx_E1, .., NOx_E100. I add them into the species database like so: 
```
NOx_PROP: &NOxproperties
  Formula: NOx
  Is_Advected: true
  Is_DryDep: true
  Is_Gas: true
  Is_Photolysis: true
  MW_g: 30.01 
NOx:
  << : *NOxproperties
  Background_VV: 4.0e-13
  FullName: Nitrogen Oxides
NOx_E1:
  << : *NOxproperties
  Background_VV: 4.0e-13
  FullName: Nitrogen Oxides. 
```
And I introduce emissions as each ensemble member in HEMCO_Config.rc. What I have found is that, the species called NOx is decaying as expected, but the ensemble members do not decay (concentrations are very elevated in comparison). I wasn't sure if anyone would be able to advice me on the most efficient way to implement the analogous NOx decay for each of these members.  Would I need to set this up as: 
```
NOx_E1             = Dummy             : k_Trop(6);
DummytoNOx      = NOx_E1              : k_Trop(7);
```
with: 
```
 loss = LNOx_chem*C(ind_NOx_E1)/NOx_lev
 prod = PNOx_chem*C(ind_NOx_E1)/NOx_lev
 k_Trop(6) = loss   /C(ind_NOx)
 k_Trop(7) = prod 
```
And manually add all 100 ensemble members into the source code, or is there a more efficient way to do this?
Thanks a lot!

**Comment by yantosca** [comment]
Also you might need to pass in NOx_lev as an argument as well, depending on how you have your code set up.

**Comment by drewpendergrass** [comment]
Hi Chloe! I don't have much to add to what Bob wrote, other than that in my EnKF code I run multiple GC simulations simultaneously (rather than one sim with multiple tracers). Sounds like your method is more computationally efficient, but you're welcome to check out my code if it helps: https://github.com/drewpendergrass/CHEEREIO

**Comment by cschooling** [comment]
Hi both, thanks so much for the fast responses. 
@yantosca, creating a function like this is a good idea! 

I have started by just trying to introduce the decay for the first ensemble member NOx_E1 to get it working, then will extend to all 100. 

However, I have added everything I remember doing for the NOx species (as described above in KPP mechanisms), but NOx_E1 concentrations have not changed to reflect that it is decaying (relative to the other ensemble members). The species has been given a KppSpcId (I can see in the log file), and if I write out the values for k_Trop(6), k_Trop(7) (current holdres for the NOx_E1 loss and production rates) then I see that these are non zero values. Is there anything I am missing here?

Btw NOx_lev, LNOx_chem, PNOx_chem are all fields I read in through HEMC.

Thanks @drewpendergrass for sharing your code. I hope I can get it working with all members in a single run for efficiency but I will have a look at how you run yours too.

**Comment by yantosca** [comment]
@cschooling: Can you attach your .kpp and .eqn files to this issue?  I can take a look.

**Comment by cschooling** [comment]
Thanks! Attached. 

[Archive.zip](https://github.com/user-attachments/files/19707405/Archive.zip)

**Comment by cschooling** [comment]
Ah ok so when I try and compile I get the following error:
```
carbon_Rates.F90:437:22:

  437 |   RCONST(7) = (k_Trop(6))
      |                      1
Warning: Array reference at (1) is out of bounds (6 > 5) in dimension 1
carbon_Rates.F90:438:22:

  438 |   RCONST(8) = (k_Trop(7))
      |                      1
Warning: Array reference at (1) is out of bounds (7 > 5) in dimension 1
gfortran -cpp -O -g  -c carbon_Monitor.F90
gfortran -cpp -O -g  -c carbon_Util.F90
gfortran -cpp -O -g  -c carbon_LinearAlgebra.F90
carbon_LinearAlgebra.F90:290:46:

  290 |       IF (incX .EQ. incY) IF (incX-1) 5,20,60
      |                                              1
Warning: Fortran 2018 deleted feature: Arithmetic IF statement at (1)
```
Could this explain why the decays havent been triggered. Do you know how I extend the dimension?

**Comment by yantosca** [comment]
Hi @cschooling, I think k_Trop is defined in commonIncludeVars.H.  You can bump up the number of elements there.

**Comment by yantosca** [comment]
And yes, if there is an out-of-bounds error that could have overwritten/corrupted data so it's possible that is the problem.

**Comment by cschooling** [comment]
Ok thanks, however I added more elements but the same issue persists. It's strange because the NOx species is definitely decaying but NOx_E1 is not and I can't find any differences relating to the two anywhere in the source code.

I was not fully sure what I should change in carbon.def to print out values. Is the aim to print out the chemistry rate values? Currently when I run ./carbon.exe the output is: 
`   NaN%. T=0.000E+00  NOx= 0.0000E+00; NOx_E1= 0.0000E+00;  `
But I just kept these lines in the script: 
```
#LOOKATALL                      { Output all species to carbon.dat}
#MONITOR NOX;NOX_E1;NOX_E2;  { Print selected species to screen        }
```
Any other advice to figure out what differences in the KPP mechanism applied to NOx compared to NOx_E1?

**Comment by yantosca** [comment]
@cschooling can you send me a zip file with your box model code?  I'll see if I can play with it a bit.

**Comment by cschooling** [comment]
Hi @yantosca just wondering if you had a chance to have a look. Still struggling to make progress on this. Thanks a lot!

**Comment by yantosca** [comment]
Hi @cschooling, I got pulled in a few different directions recently.  Will try to look at this as soon as possible, but it may take a while.

**Comment by yantosca** [comment]
Hi @cschooling, thanks for your patience.  I finally was able to take a look at your box model (and found a couple of bugs in KPP along the way, that I'm fixing now, I'll post them here for your reference once I finish the PRs).  But I was able to get the NOx_E1 and NOX_E2 to vary with time.  See  [box_mod_fixed.tar.gz](https://github.com/user-attachments/files/20041606/box_mod_fixed.tar.gz)
```console
$ make -f Makefile_carbon
$ ./carbon.exe 
   0.0%. T=0.000E+00  NOx_E2= 0.0000E+00; NOx= 0.1221E+11; NOx_E1= 0.0000E+00; 
   4.2%. T=0.360E+04  NOx_E2= 0.9109E-09; NOx= 0.1221E+11; NOx_E1= 0.9109E-09; 
   8.3%. T=0.720E+04  NOx_E2= 0.1822E-08; NOx= 0.1221E+11; NOx_E1= 0.1822E-08; 
  12.5%. T=0.108E+05  NOx_E2= 0.2733E-08; NOx= 0.1221E+11; NOx_E1= 0.2733E-08; 
  16.7%. T=0.144E+05  NOx_E2= 0.3644E-08; NOx= 0.1221E+11; NOx_E1= 0.3644E-08; 
  20.8%. T=0.180E+05  NOx_E2= 0.4554E-08; NOx= 0.1221E+11; NOx_E1= 0.4554E-08; 
  25.0%. T=0.216E+05  NOx_E2= 0.5465E-08; NOx= 0.1221E+11; NOx_E1= 0.5465E-08; 
  29.2%. T=0.252E+05  NOx_E2= 0.6376E-08; NOx= 0.1221E+11; NOx_E1= 0.6376E-08; 
  33.3%. T=0.288E+05  NOx_E2= 0.7287E-08; NOx= 0.1221E+11; NOx_E1= 0.7287E-08; 
  37.5%. T=0.324E+05  NOx_E2= 0.8198E-08; NOx= 0.1221E+11; NOx_E1= 0.8198E-08; 
  41.7%. T=0.360E+05  NOx_E2= 0.9109E-08; NOx= 0.1221E+11; NOx_E1= 0.9109E-08; 
  45.8%. T=0.396E+05  NOx_E2= 0.1002E-07; NOx= 0.1221E+11; NOx_E1= 0.1002E-07; 
  50.0%. T=0.432E+05  NOx_E2= 0.1093E-07; NOx= 0.1221E+11; NOx_E1= 0.1093E-07; 
  54.2%. T=0.468E+05  NOx_E2= 0.1184E-07; NOx= 0.1221E+11; NOx_E1= 0.1184E-07; 
  58.3%. T=0.504E+05  NOx_E2= 0.1275E-07; NOx= 0.1221E+11; NOx_E1= 0.1275E-07; 
  62.5%. T=0.540E+05  NOx_E2= 0.1366E-07; NOx= 0.1221E+11; NOx_E1= 0.1366E-07; 
  66.7%. T=0.576E+05  NOx_E2= 0.1457E-07; NOx= 0.1221E+11; NOx_E1= 0.1457E-07; 
  70.8%. T=0.612E+05  NOx_E2= 0.1549E-07; NOx= 0.1221E+11; NOx_E1= 0.1549E-07; 
  75.0%. T=0.648E+05  NOx_E2= 0.1640E-07; NOx= 0.1221E+11; NOx_E1= 0.1640E-07; 
  79.2%. T=0.684E+05  NOx_E2= 0.1731E-07; NOx= 0.1221E+11; NOx_E1= 0.1731E-07; 
  83.3%. T=0.720E+05  NOx_E2= 0.1822E-07; NOx= 0.1221E+11; NOx_E1= 0.1822E-07; 
  87.5%. T=0.756E+05  NOx_E2= 0.1913E-07; NOx= 0.1221E+11; NOx_E1= 0.1913E-07; 
  91.7%. T=0.792E+05  NOx_E2= 0.2004E-07; NOx= 0.1221E+11; NOx_E1= 0.2004E-07; 
  95.8%. T=0.828E+05  NOx_E2= 0.2095E-07; NOx= 0.1221E+11; NOx_E1= 0.2095E-07; 
 100.0%. T=0.864E+05  NOx_E2= 0.2186E-07; NOx= 0.1221E+11; NOx_E1= 0.2186E-07; 
```

One of the bus is that the Forward Euler integrator, that it wasn't returning the ending time as an output argument.   So in the main program it was never updating the simulation time.  Will push a fix to KPP and try to release a patch version ASAP.

**Comment by yantosca** [comment]
@cschooling: For your reference see these issues on the KPP repository.  I found these bugs while looking into this issue.

- https://github.com/KineticPreProcessor/KPP/issues/135
- https://github.com/KineticPreProcessor/KPP/issues/133

**Comment by cschooling** [comment]
Thanks for looking into this @yantosca!
For setting up my runs can I just use these updated .F90 scripts within the carbon KPP, or is it worth waiting for the patch version release?

**Comment by yantosca** [comment]
Hi @cschooling, KPP 3.2.1 just released today.  See: https://github.com/KineticPreProcessor/KPP/releases/tag/3.2.1

(P.S. I'm fortunate to also be a KPP developer, so we can bring in fixes like this.)

**Comment by cschooling** [comment]
Hi @yantosca, that's great thanks. I have updated my model with this new release and recompiled the carbon kpp mechanism:
`$ ./build_mechanism.sh carbon/`
However, I am still struggling with the NOx species seeing a chemical decay, but the ensemble member NOx_E1 does not. I have attached the carbon directory within KPP after recompiling. Would you expect the new release to have fixed my issue? I wonder if I am missing a step? 

[carbon.zip](https://github.com/user-attachments/files/20086347/carbon.zip)

**Comment by yantosca** [comment]
Thanks @cschooling, I'll take a look.  Thanks for your patience with this.

**Comment by yantosca** [comment]
@cschooling, can you send me a zip of your entire code dir?  Thanks.

**Comment by cschooling** [comment]
Thanks @yantosca! See attached.

**Comment by cschooling** [comment]
Hmm full code directory is possibly too large to attach. Here is the src sub directory - I havent edited anything outside of here [src.zip](https://github.com/user-attachments/files/20129185/src.zip)

**Comment by yantosca** [comment]
Thanks Chloe, glad to hear it's working!  Sorry for the delays in getting back to you, I've been juggling a few things in these days.  I'll close out this issue as resolved but do feel free to reopen it if something's still off on your end.

P.S. I will be at GCE3 in London this year and look forward to meeting the UK/European GC users in person.  Maybe see you & the Edinburgh group then.
